### PR TITLE
Ledger Tezos Validator

### PR DESCRIPTION
**Pull request by lvndry**
### 📝 Description

A Ledger [Tezos Validator](https://tzkt.io/tz3LV9aGKHDnAZHCtC9SjNtTrKRu678FqSki/operations/) is now available.
This PR adds the validators to the list of tezos validators

### ❓ Context

- **Impacted projects**: `LLC` `LLD` `LLM`
- **Linked resource(s)**: `https://ledgerhq.atlassian.net/browse/LIVE-9306` 

### ✅ Checklist

- [x] **Test coverage** 
- [x] **Atomic delivery** 
- [x] **No breaking changes** 

### 🚀 Expectations to reach

_Please make sure you follow these [**Important Steps**](https://github.com/LedgerHQ/ledger-live/blob/develop/CONTRIBUTING.md#important-steps)._

_Pull Requests must pass the CI and be internally validated in order to be merged._

### CODE REVIEW

File: apps/ledger-live-desktop/src/renderer/families/tezos/DelegateFlowModal/Body.tsx [hunk 1]
```diff
@@ -107,15 +93,15 @@ const createSteps = (params: Data): St[] => [
...
+
+const Body = ({ stepId, params, onChangeStepId, onClose }: Props) => {
   const { t } = useTranslation();
   const dispatch = useDispatch();
   const device = useSelector(getCurrentDevice);
-  const accounts = useSelector(delegatableAccountsSelector);
   const openedFromAccount = !!params.account;
-  const bakers = useBakers(whitelist);
-  const randomBaker = useRandomBaker(bakers);
-  const [steps] = useState(() => createSteps(params));
```
Comment (hedi-edelbloute) [inline apps/ledger-live-desktop/src/renderer/families/tezos/DelegateFlowModal/Body.tsx:118]: I'm not very confident in react, why did we remove this useState ?

File: libs/ledger-live-common/src/families/tezos/bakers.ts [hunk 2]
```diff
@@ -31,79 +31,106 @@ export type Delegation = {
...
+  estimatedRoi: number;
+  serviceType: string;
+  serviceHealth: string;
+  payoutTiming: string;
+  payoutAccuracy: string;
+  audit: string;
+  insuranceCoveragenumber: number;
+};
+
+const getLedgerValidator = async (): Promise<Baker> => {
```
Comment (hedi-edelbloute) [inline libs/ledger-live-common/src/families/tezos/bakers.ts:58]: getLedgerBaker ? as we use baker everywhere else

File: apps/ledger-live-desktop/src/renderer/families/tezos/DelegateFlowModal/Body.tsx [hunk 3]
```diff
@@ -27,24 +24,13 @@ import StepSummary, { StepSummaryFooter } from "./steps/StepSummary";
 import StepValidator from "./steps/StepValidator";
 import StepCustom, { StepCustomFooter } from "./steps/StepCustom";
 import StepConfirmation, { StepConfirmationFooter } from "./steps/StepConfirmation";
-import { StepId, St } from "./types";
+import { StepId, Step } from "./types";
```
Comment (hedi-edelbloute) [inline apps/ledger-live-desktop/src/renderer/families/tezos/DelegateFlowModal/Body.tsx:27]: What is different between St and Step

File: apps/ledger-live-desktop/src/renderer/families/tezos/DelegateFlowModal/steps/StepSummary.tsx [hunk 4]
```diff
@@ -162,12 +161,6 @@ const StepSummary = ({
                   <Placeholder />
                 )}
               </Container>
-
-              {isRandomChoice ? (
-                <Text ff="Inter|Medium" color="palette.text.shade60" fontSize={2}>
-                  <Trans i18nKey="delegation.flow.steps.summary.randomly" />
```
Comment (hedi-edelbloute) [inline apps/ledger-live-desktop/src/renderer/families/tezos/DelegateFlowModal/steps/StepSummary.tsx:168]: We should remove the key if it isn't used anywhere else, i didn't check

File: apps/ledger-live-mobile/src/families/tezos/DelegationFlow/Summary.tsx [hunk 5]
```diff
@@ -108,8 +103,7 @@ export default function DelegationSummary({ navigation, route }: Props) {
   const { colors } = useTheme();
   const { account, parentAccount } = useSelector(accountScreenSelector(route));
   const { t } = useTranslation();
-  const bakers = useBakers(whitelist);
-  const randomBaker = useRandomBaker(bakers);
+  const [defaultBaker] = useBakers(whitelist);
```
Comment (hedi-edelbloute) [inline apps/ledger-live-mobile/src/families/tezos/DelegationFlow/Summary.tsx:106]: ledgerBaker ? we use ledgerValidator in the codebase

File: libs/ledger-live-common/src/families/tezos/bakers.ts [hunk 6]
```diff
@@ -31,79 +31,106 @@ export type Delegation = {
...
+  const { data } = await network<API_BAKER>({
```
Comment (hedi-edelbloute) [inline libs/ledger-live-common/src/families/tezos/bakers.ts:59]: Do we really need to do this call ? Don't we do it for other bakers ?

File: libs/ledger-live-common/src/families/tezos/bakers.ts [hunk 7]
```diff
@@ -132,28 +159,32 @@ export function getAccountDelegationSync(account: AccountLike): Delegation | nul
...
+
 export function isAccountDelegating(account: AccountLike): boolean {
   return !!getAccountDelegationSync(account);
 }
-export async function loadBaker(addr: string): Promise<Baker | null | undefined> {
+
+export async function loadBaker(addr: string): Promise<Baker | undefined> {
   const cacheBaker = getBakerSync(addr);
-  if (cacheBaker) return Promise.resolve(cacheBaker);
+  if (cacheBaker) return cacheBaker;
```
Comment (hedi-edelbloute) [inline libs/ledger-live-common/src/families/tezos/bakers.ts:169]: In that case we don't return a promise right ? so the return type is Promise | Baker ?

File: apps/ledger-live-desktop/src/renderer/families/tezos/DelegateFlowModal/Body.tsx [hunk 8]
```diff
@@ -72,14 +58,14 @@ const createSteps = (params: Data): St[] => [
     label: <Trans i18nKey="delegation.flow.steps.account.label" />,
     component: StepAccount,
     footer: StepAccountFooter,
-    excludeFromBreadcrumb: Boolean(params && params.account),
+    excludeFromBreadcrumb: Boolean(params.account),
```
[the remark this answers is not in the document]
Comment (lvndry) [reply apps/ledger-live-desktop/src/renderer/families/tezos/DelegateFlowModal/Body.tsx:61]: https://github.com/LedgerHQ/ledger-live/pull/4675/files/e4d842a2d4f0f8c74a882947b0865785ab628a72#diff-6307d6870883f50f2340b08eec6007ce206d7612d15c162467432469ca527bd1R47

I you look line 47, the type data is mandatory so it's never undefined

File: apps/ledger-live-desktop/src/renderer/families/tezos/DelegateFlowModal/Body.tsx (hunk 1 above)
Comment (lvndry) [reply apps/ledger-live-desktop/src/renderer/families/tezos/DelegateFlowModal/Body.tsx:118]: useState is used to manage a state with react

`const [state, setState] = useState(() => defaultValue)`
If you don't need to update the state with a `setState` then the value never changes so it's juste a constant

File: libs/ledger-live-common/src/families/tezos/bakers.ts (hunk 7 above)
Comment (lvndry) [reply libs/ledger-live-common/src/families/tezos/bakers.ts:169]: Actually if you have an `async` function whatever you return inside this function is a Promise

File: apps/ledger-live-desktop/src/renderer/families/tezos/DelegateFlowModal/Body.tsx [hunk 9]
```diff
@@ -107,15 +93,15 @@ const createSteps = (params: Data): St[] => [
...
+const Body = ({ stepId, params, onChangeStepId, onClose }: Props) => {
   const { t } = useTranslation();
   const dispatch = useDispatch();
   const device = useSelector(getCurrentDevice);
-  const accounts = useSelector(delegatableAccountsSelector);
   const openedFromAccount = !!params.account;
-  const bakers = useBakers(whitelist);
-  const randomBaker = useRandomBaker(bakers);
-  const [steps] = useState(() => createSteps(params));
+  const [defaultBaker] = useBakers(whitelist);
```
[the remark this answers is not in the document]
Comment (lvndry) [reply apps/ledger-live-desktop/src/renderer/families/tezos/DelegateFlowModal/Body.tsx:102]: Yeah could be, I feel like it's a implementation detail, the frontend doesn't care if the defaultBaker is ledger or not, the default Baker could change if ledger decides to sunset the baker or whatever

File: apps/ledger-live-desktop/src/renderer/families/tezos/DelegateFlowModal/Body.tsx (hunk 3 above)
Comment (lvndry) [reply apps/ledger-live-desktop/src/renderer/families/tezos/DelegateFlowModal/Body.tsx:27]: No diff. Just makes the name clearer.

File: apps/ledger-live-desktop/src/renderer/families/tezos/DelegateFlowModal/steps/StepSummary.tsx (hunk 4 above)
Comment (lvndry) [reply apps/ledger-live-desktop/src/renderer/families/tezos/DelegateFlowModal/steps/StepSummary.tsx:168]: It's used

File: libs/ledger-live-common/src/families/tezos/bakers.ts (hunk 6 above)
Comment (hedi-edelbloute) [reply libs/ledger-live-common/src/families/tezos/bakers.ts]: Please do not resolve my comment without any answer

I can find the ledger validator here on this request https://tezos-bakers.api.live.ledger.com/v2/bakers

``` 
...{
    "address": "tz3LV9aGKHDnAZHCtC9SjNtTrKRu678FqSki",
    "name": "Ledger Live by Kiln",
    "logo": "https://services.tzkt.io/v1/avatars/tz3LV9aGKHDnAZHCtC9SjNtTrKRu678FqSki",
    "balance": 708166.216593,
    "stakingBalance": 708166.216593,
    "stakingCapacity": 7081662.16593,
    "maxStakingBalance": 7081662.16593,
    "freeSpace": 6373495.949337,
    "fee": 0.12,
    "minDelegation": 1,
    "payoutDelay": 1,
    "payoutPeriod": 1,
    "openForDelegation": true,
    "estimatedRoi": 0.053786,
    "serviceType": "tezos_only",
    "serviceHealth": "active",
    "payoutTiming": "no_data",
    "payoutAccuracy": "no_data",
    "insuranceCoverage": 0
  },...
  ```

Why the need to call twice the service with a more specific request
https://tezos-bakers.api.live.ledger.com/v2/bakers/tz3LV9aGKHDnAZHCtC9SjNtTrKRu678FqSki

  ```
{
  "address": "tz3LV9aGKHDnAZHCtC9SjNtTrKRu678FqSki",
  "name": "Ledger Live by Kiln",
  "logo": "https://services.tzkt.io/v1/avatars/tz3LV9aGKHDnAZHCtC9SjNtTrKRu678FqSki",
  "balance": 708186.092993,
  "stakingBalance": 708186.092993,
  "stakingCapacity": 7081860.92993,
  "maxStakingBalance": 7081860.92993,
  "freeSpace": 6373674.836936999,
  "fee": 0.12,
  "minDelegation": 1,
  "payoutDelay": 1,
  "payoutPeriod": 1,
  "openForDelegation": true,
  "estimatedRoi": 0.053787,
  "serviceType": "tezos_only",
  "serviceHealth": "active",
  "payoutTiming": "no_data",
  "payoutAccuracy": "no_data",
  "insuranceCoverage": 0,
}
```

**Comment by hedi-edelbloute** [comment]
It would be great if we do unit tests

File: libs/ledger-live-common/src/families/tezos/bakers.test.ts [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+  });
+
+  it("Ledger Baker should be first in the list", async () => {
+    const bakers = await cache();
+    expect(bakers[0].name).toBe("Ledger Live by Kiln");
+  });
+
+  it("Should convert api baker to LL baker (capacity status full)", () => {
+    const testBaker = data.find(baker => baker.freeSpace < 0);
+    if (testBaker) {
```
Comment (hedi-edelbloute) [inline libs/ledger-live-common/src/families/tezos/bakers.test.ts:119]: The test should not be conditioned

**Comment by hedi-edelbloute** [comment]
typecheck error

File: libs/ledger-live-common/src/families/tezos/bakers.test.ts [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+      headers: {},
+      statusText: "",
+      config: {},
+    };
+
+    mockedNetwork.mockReturnValue(Promise.resolve(response));
+  });
+
+  it("Ledger Baker should be first in the list", async () => {
+    const bakers = await cache();
```
Comment (hedi-edelbloute) [inline libs/ledger-live-common/src/families/tezos/bakers.test.ts:113]: We need to separate each method tested by a different describe

Comment (hedi-edelbloute) [inline libs/ledger-live-common/src/families/tezos/bakers.test.ts:113]: You should reset the LRU cache after each test as it has been done in the codebase

File: libs/ledger-live-common/src/families/tezos/bakers.test.ts (hunk 10 above)
Comment (lvndry) [reply libs/ledger-live-common/src/families/tezos/bakers.test.ts]: it's not actually in `data` there's always a baker with `freespace < 0` but `.find` always return `T | undefined`.

**Review by haammar-ledger — APPROVED** [review]
:raised_hands:

File: libs/ledger-live-common/src/families/tezos/bakers.test.ts (hunk 10 above)
Comment (hedi-edelbloute) [reply libs/ledger-live-common/src/families/tezos/bakers.test.ts]: Then it should be casted, not conditioned, as the expect may not be executed

File: libs/ledger-live-common/src/apps/support.ts [hunk 12]
```diff
@@ -35,6 +35,7 @@ const appVersionsRequired = {
   "Cardano ADA": ">= 4.1.0",
   Zcash: "> 2.0.6",
   NEAR: ">= 1.2.1",
+  "Tezos Wallet": ">= 2.4.3",
```
Comment (haammar-ledger) [inline libs/ledger-live-common/src/apps/support.ts:38]: @lvndry In the end the app that landed on P1 is **2.4.5**, so let's change this just to be safe

**Pull request merged**